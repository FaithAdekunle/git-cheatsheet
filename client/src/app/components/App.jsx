import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Categories from './categories/categories';
import Loader from './common/loader';
import Sidebar from './sidebar';
import expandOrCollapseSidebar from '../actions/sidebarAction';

export class App extends React.Component {
  constructor(props) {
    super(props);
    this.collapseSidebar = this.collapseSidebar.bind(this);
  }

  collapseSidebar() {
    const { collapseSidebar } = this.props;
    collapseSidebar();
  }

  render() {
    const { ajaxCallsInProgress } = this.props;
    return (
      <React.Fragment>
        {
          ajaxCallsInProgress ? <Loader /> : ''
        }
        <Sidebar />
        <div className="main" onClick={this.collapseSidebar}>
          <Categories />
        </div>
      </React.Fragment>
    );
  }
}

App.propTypes = {
  ajaxCallsInProgress: PropTypes.number.isRequired,
  collapseSidebar: PropTypes.func.isRequired,
};

export const mapStateToProps = state => ({
  ajaxCallsInProgress: state.ajaxCallsInProgress,
});

export const mapDispatchToProps = dispatch => ({
  collapseSidebar: () => dispatch(expandOrCollapseSidebar(false)),
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
