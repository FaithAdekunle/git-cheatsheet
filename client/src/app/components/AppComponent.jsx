import React from 'react';

const AppComponent = () => (
  <h1>Welcome to Faithâ€™s React and Express Application</h1>
);
export default AppComponent;
