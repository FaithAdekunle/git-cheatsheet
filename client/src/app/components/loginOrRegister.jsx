import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import UserActions from '../actions/userActions';

export class LoginOrRegister extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      login: true,
      register: false,
      errorResponse: '',
      user: { email: '', password: '' },
      errors: { email: '', password: '' },
    };
    this.selectLogin = this.selectLogin.bind(this);
    this.selectRegister = this.selectRegister.bind(this);
    this.onChange = this.onChange.bind(this);
    this.validateUser = this.validateUser.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  onChange(event) {
    const { target } = event;
    const { user, errors } = this.state;
    user[target.name] = target.value;
    if (!target.value) {
      errors[target.name] = `${target.name} must be provided`;
    } else {
      errors[target.name] = '';
    }
    this.setState({ user, errors });
  }

  async onSubmit(event) {
    const { user, login } = this.state;
    const { logUserIn, registerUser } = this.props;
    event.preventDefault();
    if (this.validateUser()) {
      this.setState({ errorResponse: '' });
      try {
        login ? await logUserIn(user) : await registerUser(user);
      } catch (error) {
        this.setState({ errorResponse: error.message });
      }
    }
  }

  selectLogin() {
    this.setState({ login: true, register: false, errorResponse: '' });
  }

  selectRegister() {
    this.setState({ login: false, register: true, errorResponse: '' });
  }

  validateUser() {
    const { user, errors } = this.state;
    const validateEmail = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    if (!user.password) {
      errors.password = 'password must be provided';
    } else {
      errors.password = '';
    }
    if (!user.email) {
      errors.email = 'email must be provided';
    } else if (!validateEmail.test(user.email.toLowerCase())) {
      errors.email = 'email must be valid';
    } else {
      errors.email = '';
    }
    this.setState({ errors });
    return !(errors.email || errors.password);
  }

  render() {
    const {
      login,
      register,
      user,
      errors,
      errorResponse,
    } = this.state;
    return (
      <div className="text-center">
        <span
          className={`authenticate login${login ? ' active' : ''}`}
          onClick={this.selectLogin}
        >
          <b>Login</b>
        </span>
        {' | '}
        <span
          className={`authenticate register${register ? ' active' : ''}`}
          onClick={this.selectRegister}
        >
          <b>Register</b>
        </span>
        <div className="card text-left">
          <div className="card-body">
            {
              errorResponse ? <div className="error-response text-center">{errorResponse}</div> : ''
            }
            <form onSubmit={this.onSubmit}>
              <div className="form-group">
                <label>Email</label>
                <span className="error">{errors.email}</span>
                <input
                  className="form-control email-field"
                  name="email"
                  value={user.email}
                  onChange={this.onChange}
                />
              </div>
              <div className="form-group">
                <label>Password</label>
                <span className="error">{errors.password}</span>
                <input
                  className="form-control password-field"
                  type="password"
                  name="password"
                  value={user.password}
                  onChange={this.onChange}
                />
              </div>
              <div className="text-right">
                <input className="submit" type="submit" value={`${login ? 'Login' : 'Register'}`} />
              </div>
            </form>
          </div>
        </div>
      </div>
    );
  }
}

LoginOrRegister.propTypes = {
  logUserIn: PropTypes.func.isRequired,
  registerUser: PropTypes.func.isRequired,
};

export const mapStateToProps = () => ({});

export const mapDispatchToProps = dispatch => ({
  logUserIn: credentials => dispatch(UserActions.login(credentials)),
  registerUser: credentials => dispatch(UserActions.register(credentials)),
});

export default connect(mapStateToProps, mapDispatchToProps)(LoginOrRegister);
