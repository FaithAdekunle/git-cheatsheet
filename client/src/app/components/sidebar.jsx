import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import expandOrCollapseSidebar from '../actions/sidebarAction';
import LoginOrRegister from './loginOrRegister';
import UserActions from "../actions/userActions";

export class Sidebar extends React.Component {
  constructor(props) {
    super(props);
    this.expandSidebar = this.expandSidebar.bind(this);
  }

  expandSidebar() {
    const { expandSidebar } = this.props;
    expandSidebar();
  }

  render() {
    const { expand, user, logout } = this.props;
    return (
      <React.Fragment>
        <h6 className="expand-sidebar" onClick={this.expandSidebar}>>>></h6>
        <div className={`sidebar${expand ? ' sidebar-expand' : ''}`}>
          <div>
            {
              user.id ? (
                <div className="logout-container text-center">
                  <button className="logout" onClick={logout}>logout</button>
                </div>
              ) : (
                <div className="login-or-register">
                  <LoginOrRegister />
                </div>
              )
            }
            <hr />
            <p className="text-center">
              {'"Create/Edit/Delete your own cheat categories or mark them as public/private"'}
            </p>
            <hr />
            <p className="text-center">
              {'"Edit/Delete a command or add multiple commands to any of your cheat categories"'}
            </p>
            <hr />
            <p className="text-center">
              {'"Click on a cheat command to copy it to your clipboard"'}
            </p>
            <hr />
            <p className="text-center">
              {'"Find keyword related commands using the keyword search bar"'}
            </p>
            <hr />
          </div>
          <h6 className="text-center">&copy; Faith Adekunle</h6>
        </div>
      </React.Fragment>
    );
  }
}

Sidebar.propTypes = {
  expand: PropTypes.bool.isRequired,
  user: PropTypes.objectOf(PropTypes.string).isRequired,
  expandSidebar: PropTypes.func.isRequired,
  logout: PropTypes.func.isRequired,
};

export const mapStateToProps = state => ({
  expand: state.expandSidebar,
  user: state.user,
});

export const mapDispatchToProps = dispatch => ({
  expandSidebar: () => dispatch(expandOrCollapseSidebar(true)),
  logout: () => dispatch(UserActions.logout()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Sidebar);
