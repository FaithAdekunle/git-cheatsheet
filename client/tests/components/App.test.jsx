import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { App, mapStateToProps } from '../../src/app/components/App';
import mockState from '../mockState';

describe('<App />', () => {
  test('should match snapshot when there are no ajaxCallsInProgress', () => {
    const wrapper = shallow(<App ajaxCallsInProgress={0} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should match snapshot when there are ajaxCallsInProgress', () => {
    const wrapper = shallow(<App ajaxCallsInProgress={1} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should return required state properties', () => {
    expect(mapStateToProps(mockState))
      .toEqual({ ajaxCallsInProgress: mockState.ajaxCallsInProgress });
  });
});
