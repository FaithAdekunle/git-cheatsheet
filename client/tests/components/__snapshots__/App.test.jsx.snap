// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match snapshot when there are ajaxCallsInProgress 1`] = `
<Fragment>
  <Loader />
  <Connect(Categories) />
</Fragment>
`;

exports[`<App /> should match snapshot when there are no ajaxCallsInProgress 1`] = `
<Fragment>
  <Connect(Categories) />
</Fragment>
`;
